File: frontend/src/api/digitalTwin.ts
import type {
  DigitalTwinAssetsResponse,
  DigitalTwinReplayGenerateResult,
  DigitalTwinReplaySummary,
  DigitalTwinState,
} from '../types/digitalTwin'

function apiBaseUrl(): string {
  const configured = import.meta.env.VITE_BACKEND_API_URL as string | undefined
  if (configured) return configured.replace(/\/$/, '')
  if (window.location.port && window.location.port !== '5173') return window.location.origin
  return `${window.location.protocol}//${window.location.hostname}:8000`
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${apiBaseUrl()}${path}`, {
    headers: { 'Content-Type': 'application/json', ...init?.headers },
    ...init,
  })
  const body = await response.json()
  if (!response.ok) throw new Error(JSON.stringify(body))
  return body as T
}

export function fetchDigitalTwinState(): Promise<DigitalTwinState> {
  return request<DigitalTwinState>('/api/digital-twin/state')
}

export function fetchDigitalTwinAssets(): Promise<DigitalTwinAssetsResponse> {
  return request<DigitalTwinAssetsResponse>('/api/digital-twin/assets')
}

export function generateDigitalTwinReplay(): Promise<DigitalTwinReplayGenerateResult> {
  return request<DigitalTwinReplayGenerateResult>('/api/digital-twin/replay/generate', { method: 'POST' })
}

export function fetchLatestDigitalTwinReplay(): Promise<DigitalTwinReplaySummary> {
  return request<DigitalTwinReplaySummary>('/api/digital-twin/replay/latest')
}

export function logDigitalTwinPanelRendered(): Promise<{ accepted: boolean, no_physical_command_generated: boolean }> {
  return request<{ accepted: boolean, no_physical_command_generated: boolean }>('/api/digital-twin/panel-rendered', { method: 'POST' })
}

File: backend/app/api/digital_twin.py
from fastapi import APIRouter, Depends

from app.api.deps import get_runtime
from app.schemas.digital_twin import (
    DigitalTwinAssetsResponse,
    DigitalTwinReplayGenerateResult,
    DigitalTwinReplaySummary,
    DigitalTwinState,
)
from app.services.runtime_state import RuntimeState

router = APIRouter(prefix="/api/digital-twin", tags=["digital-twin"])


@router.get("/state", response_model=DigitalTwinState)
def get_digital_twin_state(runtime: RuntimeState = Depends(get_runtime)) -> DigitalTwinState:
    return runtime.digital_twin.state(runtime)


@router.get("/assets", response_model=DigitalTwinAssetsResponse)
def get_digital_twin_assets(runtime: RuntimeState = Depends(get_runtime)) -> DigitalTwinAssetsResponse:
    return runtime.digital_twin.assets()


@router.get("/replay/latest", response_model=DigitalTwinReplaySummary)
def get_latest_digital_twin_replay(runtime: RuntimeState = Depends(get_runtime)) -> DigitalTwinReplaySummary:
    return runtime.digital_twin.latest_replay()


@router.get("/replay/{run_id}", response_model=DigitalTwinReplaySummary)
def get_digital_twin_replay(run_id: str, runtime: RuntimeState = Depends(get_runtime)) -> DigitalTwinReplaySummary:
    replay = runtime.digital_twin.latest_replay()
    return replay.model_copy(update={"run_id": run_id or replay.run_id})


@router.post("/replay/generate", response_model=DigitalTwinReplayGenerateResult)
def generate_digital_twin_replay(runtime: RuntimeState = Depends(get_runtime)) -> DigitalTwinReplayGenerateResult:
    return runtime.digital_twin.generate_replay_report()


@router.post("/panel-rendered")
def digital_twin_panel_rendered(runtime: RuntimeState = Depends(get_runtime)) -> dict[str, bool | str]:
    return runtime.digital_twin.panel_rendered()
